docs/design/hooks.md

```diff
@@ -0,0 +1,300 @@
+
+### Proposed Implementation
+NFS-Provisioner will use user-provided Configmap to learn hooks configuration. These hooks will be executed on volume provisioning or deleting events to set the given information on provided resources. 
+
+### High-Level Design
+User will deploy nfs-provisioner with Configmap having information about hook information. This Configmap should exist in the same namespace in which nfs-provisioner is deployed. Sample nfs-provisioner deployment config is as below:
+
+```yaml
+    spec:
+      serviceAccountName: openebs-maya-operator
+      containers:
+      - name: openebs-provisioner-nfs
+        image: openebs/provisioner-nfs:ci
+        env:
+        - name: OPENEBS_IO_HOOK_CONFIG
+          value: "nfs-hook"
+        livenessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - test `pgrep "^provisioner-nfs.*"` = 1
+          initialDelaySeconds: 30
+          periodSeconds: 60
+```
+
+User needs to provide Configmap name as a value of `OPENEBS_IO_HOOK_CONFIG` environment variable.
+
+### Low-Level Design
+#### Hook Definition
+NFS Provisioner uses `Provisioner` to provision NFS PV. The existing `Provisioner` definition needs to be extended to use the following hooks definition.
+
+```go
+// HookActionType defines type of action for annotation and finalizer
+type HookActionType string
+
+const (
+	// HookActionAdd represent add action
+	HookActionAdd HookActionType = "Add"
+	// HookActionAdd represent remove action
+	HookActionRemove HookActionType = "Remove"
+)
+
+// ProvisionerEventType defines the type of events on which hook needs to be executed
+type ProvisionerEventType string
+
+const (
+	// ProvisionerEventCreate represent create event
+	ProvisionerEventCreate ProvisionerEventType = "Create"
+	// ProvisionerEventDelete represent delete event
+	ProvisionerEventDelete ProvisionerEventType = "Delete"
+)
+
+// PVHook defines the field which will be updated for PV Hook Action
+type PVHook struct {
+	// Annotations needs to be added/removed on/from the PV
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Finalizers needs to be added/removed on/from the PV
+	Finalizers []string `json:"finalizers,omitempty"`
+}
+
+// PVCHook defines the field which will be updated for PVC Hook Action
+type PVCHook struct {
+	// Annotations needs to be added/removed on/from the PVC
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Finalizers needs to be added/removed on/from the PVC
+	Finalizers []string `json:"finalizers,omitempty"`
+}
+
+// ServiceHook defines the field which will be updated for Service Hook Action
+type ServiceHook struct {
+	// Annotations needs to be added/removed on/from the Service
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Finalizers needs to be added/removed on/from the Service
+	Finalizers []string `json:"finalizers,omitempty"`
+}
```

Comment:
nit/optional: We can have common hook and all these PVCHook, PVHook, ServiceHook can embed that common hook(I am seeing this like some metadata which will be common in all APIs).
```go
type MetaDataHook struct {
   Annotation map[string]string `json:"annotations,"`
   Finalizers []string `json:"finalizers,omitempty"`
}

type PVHook struct {
  MetaDataHook
}

type PVCHook struct {
  MetaDataHook
}
...
...
...
```

**Ignore this comment** if the intention is to have a loosely coupled nature.